File: app/src/main/java/com/example/recipeclipper/data/model/RecipeTextWindow.kt
package com.example.recipeclipper.data.model

/**
 * Which part of a page's text goes to the on-device model (#103), which reads only a few
 * thousand words. Pure; the iOS app's `RecipeTextWindow` is the same.
 *
 * The anchor is the ingredients heading ("Ingredients", "Zutaten", "材料": `headings.json`, every
 * shipped language at once) followed by the most lines that look like ingredient lines, a
 * little more if a steps heading follows. With no such heading, it is the densest run of
 * ingredient-looking lines. With neither, the page has no recipe to find: null, and the model
 * isn't asked. The window keeps a few lines before the anchor (the recipe card's title, times
 * and servings), then as many lines after it as fit, and starts with the page's title.
 */
object RecipeTextWindow {
    /** A heading is a short line. */
    private const val HEADING_MAX = 60

    /** The lines after a heading that count towards it. */
    private const val LOOK_AHEAD = 40

    /** The most lines kept before the anchor, within a fifth of the budget. */
    private const val LEAD_LINES = 12

    private const val INGREDIENT_MAX = 100
    private const val SHORT_LINE = 40

    private val headings: Pair<List<String>, List<String>> by lazy {
        val all = LanguageWords.SHIPPED.mapNotNull { LanguageWords.forTag(it) }
        all.flatMap { it.strings("headings", "ingredients") } to all.flatMap { it.strings("headings", "steps") }
    }

    /** The window's text, lines joined by "\n", at most [maxChars] long; null if no recipe shows. */
    fun window(page: PageText, maxChars: Int): String? {
        val lines = page.lines.map { it.trim() }.filter { it.isNotEmpty() }
        val anchor = anchor(lines) ?: return null
        val title = page.title?.trim()?.takeIf { it.isNotEmpty() && it.length < maxChars / 4 }
        var budget = maxChars - (title?.length?.plus(1) ?: 0)

        val lead = ArrayDeque<String>()
        var leadChars = 0
        var i = anchor - 1
        while (i >= 0 && lead.size < LEAD_LINES && leadChars + lines[i].length + 1 <= maxChars / 5) {
            lead.addFirst(lines[i]); leadChars += lines[i].length + 1; i--
        }
        budget -= leadChars
        val body = mutableListOf<String>()
        var j = anchor
        while (j < lines.size && lines[j].length + 1 <= budget) {
            body += lines[j]; budget -= lines[j].length + 1; j++
        }
        if (body.isEmpty()) return null
        val window = lead + body
        val withTitle = if (title == null || title in window) window else listOf(title) + window
        return withTitle.joinToString("\n")
    }

    /** The index of the line the recipe starts at, or null. */
    internal fun anchor(lines: List<String>): Int? {
        val looks = lines.map(::looksLikeIngredient)
        var best: Int? = null
        var bestScore = 1
        for ((i, line) in lines.withIndex()) {
            if (!isHeading(line, headings.first)) continue
            val ahead = (i + 1 until minOf(lines.size, i + 1 + LOOK_AHEAD)).count { looks[it] }
            val steps = (i + 1 until minOf(lines.size, i + 1 + 2 * LOOK_AHEAD)).any { isHeading(lines[it], headings.second) }
            val score = ahead + if (steps) 2 else 0
            if (ahead >= 1 && score > bestScore) { best = i; bestScore = score }
        }
        if (best != null) return best
        // No heading: the start of the densest run of ingredient-looking lines, three or more in ten.
        var dense: Int? = null
        var denseCount = 2
        for (i in lines.indices) {
            if (!looks[i]) continue
            val count = (i until minOf(lines.size, i + 10)).count { looks[it] }
            if (count > denseCount) { dense = i; denseCount = count }
        }
        return dense
    }

    /** A short line that starts with one of [words] as whole words ("Ingredients:", "INGREDIENTS"). */
    internal fun isHeading(line: String, words: List<String> = headings.first + headings.second): Boolean {
        if (line.length > HEADING_MAX) return false
        val text = line.lowercase().trimStart { !it.isLetterOrDigit() }
        return words.any { word ->
            text.startsWith(word) && (text.length == word.length || !text[word.length].isLetterOrDigit())
        }
    }

    /** "2 cups flour", "• ½ tsp salt", "Mehl 200 g", "砂糖 大さじ2": an amount first, or a short line with one. */
    internal fun looksLikeIngredient(line: String): Boolean {
        if (line.length > INGREDIENT_MAX) return false
        val text = line.trimStart { !it.isLetterOrDigit() && it !in VULGAR_FRACTIONS }
        val first = text.firstOrNull() ?: return false
        return first.isDigit() || first in VULGAR_FRACTIONS || (text.length <= SHORT_LINE && text.any { it.isDigit() })
    }

    private const val VULGAR_FRACTIONS = "¼½¾⅐⅑⅒⅓⅔⅕⅖⅗⅘⅙⅚⅛⅜⅝⅞"
}

File: ios/RecipeClipper/Data/Model/RecipeTextWindow.swift
import Foundation

/// Which part of a page's text goes to the on-device model (#103). Pure; Android's
/// `RecipeTextWindow`, rule for rule: the ingredients heading followed, before the next heading,
/// by the most ingredient-looking lines (more if the steps heading follows), else the densest
/// run of such lines, else nil (the model isn't asked). The window keeps a few lines before the
/// anchor, then as many after it as fit, and starts with the page's title.
enum RecipeTextWindow {
    private static let headingMax = 60
    private static let lookAhead = 40
    private static let leadLines = 12
    private static let ingredientMax = 100
    private static let shortLine = 40
    private static let vulgarFractions = Set("¼½¾⅐⅑⅒⅓⅔⅕⅖⅗⅘⅙⅚⅛⅜⅝⅞".unicodeScalars)

    private static let headings: (ingredients: [String], steps: [String]) = {
        let all = LanguageWords.shipped.compactMap { LanguageWords.forTag($0) }
        return (all.flatMap { $0.strings("headings", "ingredients") }, all.flatMap { $0.strings("headings", "steps") })
    }()

    /// The window's text, lines joined by "\n", at most `maxChars` long; nil if no recipe shows.
    static func window(_ page: PageText, maxChars: Int) -> String? {
        let lines = page.lines.map { $0.kTrimmed }.filter { !$0.isEmpty }
        guard let anchor = anchor(lines) else { return nil }
        let title = page.title.map { $0.kTrimmed }.flatMap { !$0.isEmpty && length($0) < maxChars / 4 ? $0 : nil }
        var budget = maxChars - (title.map { length($0) + 1 } ?? 0)

        var lead: [String] = []
        var leadChars = 0
        var i = anchor - 1
        while i >= 0, lead.count < leadLines, leadChars + length(lines[i]) + 1 <= maxChars / 5 {
            lead.insert(lines[i], at: 0); leadChars += length(lines[i]) + 1; i -= 1
        }
        budget -= leadChars
        var body: [String] = []
        var j = anchor
        while j < lines.count, length(lines[j]) + 1 <= budget {
            body.append(lines[j]); budget -= length(lines[j]) + 1; j += 1
        }
        if body.isEmpty { return nil }
        let window = lead + body
        let withTitle = title.map { window.contains($0) ? window : [$0] + window } ?? window
        return withTitle.joined(separator: "\n")
    }

    /// The index of the line the recipe starts at, or nil.
    static func anchor(_ lines: [String]) -> Int? {
        let looks = lines.map(looksLikeIngredient)
        var best: Int?
        var bestScore = 1
        for (i, line) in lines.enumerated() where isHeading(line, headings.ingredients) {
            var ahead = 0
            var steps = false
            var j = i + 1
            while j < lines.count, j <= i + lookAhead {
                if isHeading(lines[j], headings.steps) { steps = true; break }
                if isHeading(lines[j], headings.ingredients) { break }
                if looks[j] { ahead += 1 }
                j += 1
            }
            let score = ahead + (steps ? 2 : 0)
            if ahead >= 1, score > bestScore { best = i; bestScore = score }
        }
        if let best { return best }
        var dense: Int?
        var denseCount = 2
        for i in lines.indices where looks[i] {
            let count = (i..<min(lines.count, i + 10)).filter { looks[$0] }.count
            if count > denseCount { dense = i; denseCount = count }
        }
        return dense
    }

    /// A short line that starts with one of `words` as whole words ("Ingredients:", "INGREDIENTS").
    static func isHeading(_ line: String, _ words: [String]? = nil) -> Bool {
        if length(line) > headingMax { return false }
        let text = Array(line.lowercased().utf16).drop { !PageRecipeCheck.isLetterOrDigit($0) }
        return (words ?? headings.ingredients + headings.steps).contains { word in
            let w = Array(word.utf16)
            guard text.count >= w.count, Array(text.prefix(w.count)) == w else { return false }
            return text.count == w.count || !PageRecipeCheck.isLetterOrDigit(text[text.startIndex + w.count])
        }
    }

    /// An ingredients heading in any shipped language; `PageRecipeCheck` cuts a page into cards with it (#128).
    static func isIngredientsHeading(_ line: String) -> Bool { isHeading(line, headings.ingredients) }

    static func isStepsHeading(_ line: String) -> Bool { isHeading(line, headings.steps) }

    /// "2 cups flour", "• ½ tsp salt", "Mehl 200 g", "砂糖 大さじ2": an amount first, or a short line with one.
    static func looksLikeIngredient(_ line: String) -> Bool {
        if length(line) > ingredientMax { return false }
        let text = Array(line.utf16).drop { !PageRecipeCheck.isLetterOrDigit($0) && !isFraction($0) }
        guard let first = text.first else { return false }
        return PageRecipeCheck.isDigit(first) || isFraction(first)
            || (text.count <= shortLine && text.contains(where: PageRecipeCheck.isDigit))
    }

    private static func isFraction(_ c: UInt16) -> Bool { Unicode.Scalar(c).map(vulgarFractions.contains) ?? false }

    /// Length as Kotlin counts it: UTF-16 units.
    private static func length(_ s: String) -> Int { s.utf16.count }
}
